### Pandas Lab -- Cleaning, Merging, & Grouping

This lab is designed to introduce students to common use cases for Pandas when working with data:

 - Creating new information out of your existing data set
 - Merging, concatenating, and joining different data sources
 - Grouping -- With both time & non-time based data

### Section I: Creating Data Out of Your Existing Columns

Go ahead and create the following columns in your dataset.

**Column 1:**

  - **Column Name:** Profitable
  - **Values:** `True` if `Profit` > 0, `False` if not.

In [10]:
import pandas as pd
import numpy as np
sales = pd.read_excel("/Users/aoifeduna/AoifeRepo/aoiferepo/Lectures/Unit2/data/superstore.xls")

In [13]:
sales.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1       42420   South  FUR-CH-10000454        Furniture       Chairs   
2       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3       33311   South  FUR-TA-10000577        Furniture       Tables   
4       33311   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

In [14]:
sales['Profitable'] = np.where(Sales['Profit'] > 0, True, False)

NameError: name 'Sales' is not defined

In [16]:
conditions = [
    (sales['Ship Mode'] == 'Same Day'),
    (sales['Ship Mode'] == 'First Class'),
    (sales['Ship Mode'] == 'Second Class'),
    (sales['Ship Mode'] == 'Standard Class')
]

results = [0,2,3,6]
sales['Expected Ship Time'] = np.select(conditions, results, -1)

In [28]:
sales.dtypes

Row ID                         int64
Order ID                      object
Order Date            datetime64[ns]
Ship Date             datetime64[ns]
Ship Mode                     object
Customer ID                   object
Customer Name                 object
Segment                       object
Country                       object
City                          object
State                         object
Postal Code                    int64
Region                        object
Product ID                    object
Category                      object
Sub-Category                  object
Product Name                  object
Sales                        float64
Quantity                       int64
Discount                     float64
Profit                       float64
Expected Ship Time             int64
Actual Ship Time               int64
Late                            bool
dtype: object

In [17]:
sales.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ... Region  \
0      Claire Gute   Consumer  United States        Henderson  ...  South   
1      Claire Gute   Consumer  United States        Henderson  ...  South   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   West   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...  South   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...  South   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   
3  FUR-TA-10000577        Furniture       Tables   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

   Discount    Profit  Expected Ship Time  
0      0.00   41.9136                   3  
1      0.00  219.5820                   3  
2      0.00    6.8714                   3  
3      0.45 -383.0310                   6  
4      0.20    2.5164                   6  

[5 rows x 22 columns]

**Column 3:**

 - **Column Name:** Actual Ship Time
 - **Values:**
   - `Ship Date` - `Order Date`
 - **Note:** When you subtract these columns, your column will be a **time delta**.  See if you can use the `dt` attribute to convert these values into an integer.  Ie, if your value reads `3 days`, you want that to be 3 instead.  You can read more about different time periods in pandas here:  https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#time-date-components

In [29]:
sales['Actual Ship Time'] = (sales['Ship Date'] - sales['Order Date']).dt.days
# dt stands for date time
# Tons of different date properties you can extract

In [30]:
sales['Order Date'].dt.day

0        8
1        8
2       12
3       11
4       11
        ..
9989    21
9990    26
9991    26
9992    26
9993     4
Name: Order Date, Length: 9994, dtype: int64

In [24]:
sales.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   
3  FUR-TA-10000577        Furniture       Tables   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount    Profit  Expected Ship Time  Actual Ship Time  
0     0.00   41.9136                   3                 3  
1     0.00  219.5820                   3                 3  
2     0.00    6.8714                   3                 4  
3     0.45 -383.0310                   6                 7  
4     0.20    2.5164                   6                 7  

[5 rows x 23 columns]

**Column 4:**

 - **Column Name:** Late
 - **Values:** `True` if `Actual Ship Time` > `Expected Ship Time`, `False` otherwise

In [25]:
sales['Late'] = np.where(sales['Actual Ship Time'] > sales['Expected Ship Time'], True, False)

In [26]:
sales.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

          Category  Sub-Category  \
0        Furniture     Bookcases   
1        Furniture        Chairs   
2  Office Supplies        Labels   
3        Furniture        Tables   
4  Office Supplies       Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount    Profit  Expected Ship Time  Actual Ship Time   Late  
0     0.00   41.9136                   3                 3  False  
1     0.00  219.5820                   3                 3  False  
2     0.00    6.8714                   3                 4   True  
3     0.45 -383.0310                   6                 7   True  
4     0.20    2.5164                   6                 7   True  

[5 rows x 24 columns]

In [ ]:
# Can use apply to run functions along axes
# How you fuse traditional python programming with data cleaning

### Section II: Merging Dataframes

This excel spreadsheet has 3 separate sheets.  Look up the documentation on the `pd.read_excel` method on how to load in the other two.  

After that, merge the other two dataframes into your original one, and make sure your original dataset now has the following columns:

 - **Salesperson:** This is the Salesperson in charge of each region.
 - **Returned:** This details whether or not the order was returned.  Fill in null values with the value `no`.
 
Use the `drop()` method if you need to get rid of redundant columns.

**Important:** We want to keep all of the rows in the dataset we first loaded in.  After each merge, it's a good idea to make sure your dataset hasn't shrunk, which will happen if you don't choose the right merge type.  Make sure you have 9,994 rows when you're finished!

In [11]:
?pd.read_excel

In [107]:
import pandas as pd
df = pd.read_excel("/Users/aoifeduna/AoifeRepo/aoiferepo/Lectures/Unit2/data/superstore.xls")

df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1       42420   South  FUR-CH-10000454        Furniture       Chairs   
2       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3       33311   South  FUR-TA-10000577        Furniture       Tables   
4       33311   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

In [108]:
df2 = pd.read_excel("/Users/aoifeduna/AoifeRepo/aoiferepo/Lectures/Unit2/data/superstore.xls", sheet_name=1)
# You don't need to specify the name of the sheet, just the number.
# By default it just reads in the first sheet name.
df3 = pd.read_excel("/Users/aoifeduna/AoifeRepo/aoiferepo/Lectures/Unit2/data/superstore.xls", sheet_name=2)

In [109]:
df2.head()
# You can see the returned column
# But this is just the list of the orders that were returned

Returned        Order ID
0      Yes  CA-2017-153822
1      Yes  CA-2017-129707
2      Yes  CA-2014-152345
3      Yes  CA-2015-156440
4      Yes  US-2017-155999

In [110]:
df3.head()
# The list of sales people

Person   Region
0      Anna Andreadi     West
1        Chuck Magee     East
2     Kelly Williams  Central
3  Cassandra Brandow    South

In [111]:
df = df.merge(df2, on='Order ID', how='left')
# We specify the join key, which in this case is Order ID
# Since they're spelled the same we can just use on. If they were different we would use left on and right on.
# If you don't specify left, you will get more data. If you ordered multiple items
# they're separated out into their own line.
# Without the left join, it's joining on every single possible merge.
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ... Region  \
0      Claire Gute   Consumer  United States        Henderson  ...  South   
1      Claire Gute   Consumer  United States        Henderson  ...  South   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   West   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...  South   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...  South   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   
3  FUR-TA-10000577        Furniture       Tables   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

   Discount    Profit  Returned  
0      0.00   41.9136       NaN  
1      0.00  219.5820       NaN  
2      0.00    6.8714       NaN  
3      0.45 -383.0310       NaN  
4      0.20    2.5164       NaN  

[5 rows x 22 columns]

In [112]:
# Now, for every single product that was bought we want to have a link of the sales person that was in charge
df['Region'].value_counts()

West       3203
East       2848
Central    2323
South      1620
Name: Region, dtype: int64

In [113]:
df = df.merge(df3, on='Region', how='left')

In [114]:
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   
3  FUR-TA-10000577        Furniture       Tables   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount    Profit  Returned             Person  
0     0.00   41.9136       NaN  Cassandra Brandow  
1     0.00  219.5820       NaN  Cassandra Brandow  
2     0.00    6.8714       NaN      Anna Andreadi  
3     0.45 -383.0310       NaN  Cassandra Brandow  
4     0.20    2.5164       NaN  Cassandra Brandow  

[5 rows x 23 columns]

In [115]:
df['Returned'].fillna('No', inplace=True)

In [116]:
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   
3  FUR-TA-10000577        Furniture       Tables   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount    Profit  Returned             Person  
0     0.00   41.9136        No  Cassandra Brandow  
1     0.00  219.5820        No  Cassandra Brandow  
2     0.00    6.8714        No      Anna Andreadi  
3     0.45 -383.0310        No  Cassandra Brandow  
4     0.20    2.5164        No  Cassandra Brandow  

[5 rows x 23 columns]

### Section III: Grouping

Use the `groupby` or `resample` method to answer the following questions.

**Question 1:** What salesperson had the highest average sales amount? 

In [188]:
df.groupby('Person').agg('mean').sort_values(by='Sales', ascending=True)
# This is correct, I'm just running it much later than other things I've run

KeyError: 'Person'

In [187]:
df.groupby('Person')['Sales'].mean().idxmax()
# This is also correct, just running it much later

KeyError: 'Person'

**Question 2:** Within each ship mode, compare how likely late orders were to be profitable or not

In [118]:
import numpy as np
df['Profitable'] = np.where(df['Profit'] > 0, True, False)
df['Actual Ship Time'] = (df['Ship Date'] - df['Order Date']).dt.days
conditions = [
    df['Ship Mode'] == 'Same Day',
    df['Ship Mode'] == 'First Class',
    df['Ship Mode'] == 'Second Class',
    df['Ship Mode'] == 'Standard Class'
]

results = [0, 2, 3, 6]

df['Expected Ship Time'] = np.select(conditions, results, -1)
df['Late'] = df['Actual Ship Time'] > df['Expected Ship Time']

In [119]:
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...     Sales  \
0      Claire Gute   Consumer  United States        Henderson  ...  261.9600   
1      Claire Gute   Consumer  United States        Henderson  ...  731.9400   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   14.6200   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...  957.5775   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   22.3680   

   Quantity Discount    Profit Returned             Person Profitable  \
0         2     0.00   41.9136       No  Cassandra Brandow       True   
1         3     0.00  219.5820       No  Cassandra Brandow       True   
2         2     0.00    6.8714       No      Anna Andreadi       True   
3         5     0.45 -383.0310       No  Cassandra Brandow      False   
4         2     0.20    2.5164       No  Cassandra Brandow       True   

   Actual Ship Time  Expected Ship Time   Late  
0                 3                   3  False  
1                 3                   3  False  
2                 4                   3   True  
3                 7                   6   True  
4                 7                   6   True  

[5 rows x 27 columns]

In [124]:
df.groupby(['Ship Mode', 'Late'])['Profitable'].mean()

Ship Mode       Late 
First Class     False    0.795405
                True     0.814103
Same Day        False    0.813102
                True     0.916667
Second Class    False    0.845415
                True     0.827500
Standard Class  False    0.794651
                True     0.805153
Name: Profitable, dtype: float64

**Question 3:** What ship mode had the most consistently on time orders?

In [129]:
df.groupby('Ship Mode')['Late'].mean()

Ship Mode
First Class       0.405722
Same Day          0.044199
Second Class      0.411311
Standard Class    0.104055
Name: Late, dtype: float64

**Question 4:** For each sales person, get their average, median, max, and count of their sales.

In [135]:
df.groupby('Person')['Sales'].agg(['mean','median','max','count'])


mean  median        max  count
Person                                                 
Anna Andreadi      226.493233  60.840  13999.960   3203
Cassandra Brandow  241.803645  54.594  22638.480   1620
Chuck Magee        238.336110  54.900  11199.968   2848
Kelly Williams     215.772661  45.980  17499.950   2323

**Question 5:** Group your dataset according to  `Region`, and `Category`, and then call the `describe()` method to get the summary statistics for each subgroup.

In [136]:
df.groupby(['Region','Category']).describe()

Row ID                                           \
                          count         mean          std    min     25%   
Region  Category                                                           
Central Furniture         481.0  5198.831601  2965.518718   37.0  2788.0   
        Office Supplies  1422.0  5025.764416  2890.995980   15.0  2601.5   
        Technology        420.0  4854.019048  2980.905419   36.0  2045.0   
East    Furniture         601.0  4976.960067  2927.036861   24.0  2330.0   
        Office Supplies  1712.0  4926.139019  2971.616543   29.0  2194.5   
        Technology        535.0  4951.805607  2892.183235   48.0  2432.5   
South   Furniture         332.0  5019.783133  2848.584072    1.0  2699.5   
        Office Supplies   995.0  5105.296482  2778.947757    5.0  2693.0   
        Technology        293.0  5006.679181  2774.039447  107.0  2780.0   
West    Furniture         707.0  4999.953324  2814.077939    6.0  2614.5   
        Office Supplies  1897.0  4929.139694  2841.198338    3.0  2450.0   
        Technology        599.0  5152.409015  2910.770129    8.0  2720.5   

                                                 Postal Code                \
                            50%      75%     max       count          mean   
Region  Category                                                             
Central Furniture        5168.0  7949.00  9963.0       481.0  65486.453222   
        Office Supplies  4948.0  7584.50  9983.0      1422.0  65275.734880   
        Technology       4793.0  7373.75  9984.0       420.0  65393.026190   
East    Furniture        4826.0  7562.00  9965.0       601.0  17317.247920   
        Office Supplies  4929.0  7513.25  9986.0      1712.0  17401.548481   
        Technology       5040.0  7420.00  9947.0       535.0  18640.338318   
South   Furniture        5322.0  7405.50  9990.0       332.0  34552.117470   
        Office Supplies  5053.0  7519.50  9972.0       995.0  34243.389950   
        Technology       4912.0  7479.00  9989.0       293.0  35261.351536   
West    Furniture        5145.0  7360.00  9991.0       707.0  91680.442716   
        Office Supplies  4895.0  7319.00  9994.0      1897.0  91769.716921   
        Technology       5401.0  7669.50  9992.0       599.0  91543.449082   

                         ... Actual Ship Time      Expected Ship Time  \
                         ...              75%  max              count   
Region  Category         ...                                            
Central Furniture        ...              5.0  7.0              481.0   
        Office Supplies  ...              5.0  7.0             1422.0   
        Technology       ...              5.0  7.0              420.0   
East    Furniture        ...              5.0  7.0              601.0   
        Office Supplies  ...              5.0  7.0             1712.0   
        Technology       ...              5.0  7.0              535.0   
South   Furniture        ...              5.0  7.0              332.0   
        Office Supplies  ...              5.0  7.0              995.0   
        Technology       ...              5.0  7.0              293.0   
West    Furniture        ...              5.0  7.0              707.0   
        Office Supplies  ...              5.0  7.0             1897.0   
        Technology       ...              5.0  7.0              599.0   

                                                                      
                             mean       std  min  25%  50%  75%  max  
Region  Category                                                      
Central Furniture        4.532225  1.958821  0.0  3.0  6.0  6.0  6.0  
        Office Supplies  4.626582  1.913822  0.0  3.0  6.0  6.0  6.0  
        Technology       4.447619  1.899534  0.0  3.0  6.0  6.0  6.0  
East    Furniture        4.377704  1.982112  0.0  3.0  6.0  6.0  6.0  
        Office Supplies  4.438668  1.972867  0.0  3.0  6.0  6.0  6.0  
        Technology       4.444860  1.994432  0.0 

**Question 6:** Use the `Resample()` method to get the sum of sales for each quarter.

In [137]:
df.resample('Q', on = 'Order Date')['Sales'].sum()

Order Date
2014-03-31     74447.7960
2014-06-30     86538.7596
2014-09-30    143633.2123
2014-12-31    179627.7302
2015-03-31     68851.7386
2015-06-30     89124.1870
2015-09-30    130259.5752
2015-12-31    182297.0082
2016-03-31     93237.1810
2016-06-30    136082.3010
2016-09-30    143787.3622
2016-12-31    236098.7538
2017-03-31    123144.8602
2017-06-30    133764.3720
2017-09-30    196251.9560
2017-12-31    280054.0670
Freq: Q-DEC, Name: Sales, dtype: float64

**Question 7:** What quarter had the highest total sales amount?

In [141]:
df.resample('Q', on='Order Date')['Sales'].sum().max()

280054.06700000004

In [143]:
df.resample('Q', on='Order Date')['Sales'].sum().idxmax()
# This returns the index of the maximum value displayed above

Timestamp('2017-12-31 00:00:00', freq='Q-DEC')

**Question 8:** See if you can use the `groupby` method to get a list of yearly sales for each region inside the dataset.

**Hint:** Try using the `dt` attribute of the `Order Date` column.

In [150]:
df.groupby(['Region', df['Order Date'].dt.year])['Sales'].sum()

Region   Order Date
Central  2014          103838.1646
         2015          102874.2220
         2016          147429.3760
         2017          147098.1282
East     2014          128680.4570
         2015          156332.0570
         2016          180685.8220
         2017          213082.9040
South    2014          103845.8435
         2015           71359.9805
         2016           93610.2235
         2017          122905.8575
West     2014          147883.0330
         2015          139966.2495
         2016          187480.1765
         2017          250128.3655
Name: Sales, dtype: float64

**Bonus:** Creating summary statistics with a `groupby` statement.  

Lots of times it's very useful to be able to create a summary statistic for a particular category to compare with individual samples.

For example, if you were doing fraud detection, and someone were making a purchase at a 7-11, a transaction amount of $175 would be unusually large for someone making a purchase at a convenience store, essentially setting off a red flag that the transaction might be suspicious.

Creating such comparisons is easily done using the `groupby` method and then merging it back into the original dataframe.  

For example, if you wanted to compare every single purchase amount with the average amount for that category, you could do it in the following way:

In [176]:
import pandas as pd
df = pd.read_excel("/Users/aoifeduna/AoifeRepo/aoiferepo/Lectures/Unit2/data/superstore.xls")
# create the grouping
cat_grouping = df.groupby('Category')[['Sales']].mean()
# this step is mostly just to make the merged dataframe more tidy
cat_grouping.rename({'Sales': 'Cat_Average'}, axis=1, inplace=True)

In [177]:
# join them
df = df.merge(cat_grouping, left_on='Category', right_index=True)

In [178]:
# and now we can see each purchase amount compared to the average amt
# for that category
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0        1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1        2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
3        4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
5        6  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
10      11  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   

      Customer Name   Segment        Country             City  ... Region  \
0       Claire Gute  Consumer  United States        Henderson  ...  South   
1       Claire Gute  Consumer  United States        Henderson  ...  South   
3    Sean O'Donnell  Consumer  United States  Fort Lauderdale  ...  South   
5   Brosina Hoffman  Consumer  United States      Los Angeles  ...   West   
10  Brosina Hoffman  Consumer  United States      Los Angeles  ...   West   

         Product ID   Category Sub-Category  \
0   FUR-BO-10001798  Furniture    Bookcases   
1   FUR-CH-10000454  Furniture       Chairs   
3   FUR-TA-10000577  Furniture       Tables   
5   FUR-FU-10001487  Furniture  Furnishings   
10  FUR-TA-10001539  Furniture       Tables   

                                         Product Name      Sales Quantity  \
0                   Bush Somerset Collection Bookcase   261.9600        2   
1   Hon Deluxe Fabric Upholstered Stacking Chairs,...   731.9400        3   
3       Bretford CR4500 Series Slim Rectangular Table   957.5775        5   
5   Eldon Expressions Wood and Plastic Desk Access...    48.8600        7   
10           Chromcraft Rectangular Conference Tables  1706.1840        9   

    Discount    Profit  Cat_Average  
0       0.00   41.9136   349.834887  
1       0.00  219.5820   349.834887  
3       0.45 -383.0310   349.834887  
5       0.00   14.1694   349.834887  
10      0.20   85.3092   349.834887  

[5 rows x 22 columns]

So for instance, if wanted to ask ourselves, "which customers consistently punch above their weight when it comes to the actual items that they buy?"

We could easily do the following:

In [179]:
# turn the difference between the two columns into a percent
df['Cat Difference'] = ((df['Sales'] / df['Cat_Average']) - 1) * 100

In [180]:
# now group and sort the values
df.groupby('Customer Name')['Cat Difference'].mean().sort_values(ascending=False)

Customer Name
Mitch Willingham        845.569285
Christopher Martinez    636.313301
Andy Reiter             449.654450
Adrian Barton           418.548520
Sanjit Chand            386.906876
                           ...    
Susan Gilcrest          -91.963736
Roy Skaria              -93.762646
Lela Donovan            -95.554963
Mitch Gastineau         -96.387490
Thais Sissman           -97.974843
Name: Cat Difference, Length: 793, dtype: float64

**Your Turn:** Using a similar methodology as above, figure out the 10 customers who are the most profitable on average, when compared to the subcategory that they purchased from.

If you wanted, you could also limit this to customers who only made a minimum number of purchases as well.

In [181]:
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0        1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1        2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
3        4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
5        6  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   
10      11  CA-2014-115812 2014-06-09 2014-06-14  Standard Class    BH-11710   

      Customer Name   Segment        Country             City  ...  \
0       Claire Gute  Consumer  United States        Henderson  ...   
1       Claire Gute  Consumer  United States        Henderson  ...   
3    Sean O'Donnell  Consumer  United States  Fort Lauderdale  ...   
5   Brosina Hoffman  Consumer  United States      Los Angeles  ...   
10  Brosina Hoffman  Consumer  United States      Los Angeles  ...   

         Product ID   Category Sub-Category  \
0   FUR-BO-10001798  Furniture    Bookcases   
1   FUR-CH-10000454  Furniture       Chairs   
3   FUR-TA-10000577  Furniture       Tables   
5   FUR-FU-10001487  Furniture  Furnishings   
10  FUR-TA-10001539  Furniture       Tables   

                                         Product Name      Sales Quantity  \
0                   Bush Somerset Collection Bookcase   261.9600        2   
1   Hon Deluxe Fabric Upholstered Stacking Chairs,...   731.9400        3   
3       Bretford CR4500 Series Slim Rectangular Table   957.5775        5   
5   Eldon Expressions Wood and Plastic Desk Access...    48.8600        7   
10           Chromcraft Rectangular Conference Tables  1706.1840        9   

   Discount    Profit  Cat_Average  Cat Difference  
0      0.00   41.9136   349.834887      -25.118960  
1      0.00  219.5820   349.834887      109.224416  
3      0.45 -383.0310   349.834887      173.722701  
5      0.00   14.1694   349.834887      -86.033411  
10     0.20   85.3092   349.834887      387.711221  

[5 rows x 23 columns]

In [182]:
# create the grouping
subcat_grouping = df.groupby('Sub-Category')[['Profit']].mean()
# do some renaming to make things tidier
subcat_grouping.rename({'Profit': 'Subcat Avg'}, axis=1, inplace=True)
# merge
df = df.merge(subcat_grouping, left_on='Sub-Category', right_index=True)

In [183]:
# compare differences between the two columns, as a percentage
df['Subcat Diff'] = ((df['Profit']/df['Subcat Avg']) - 1) *100

In [184]:
# now group customers according to the column you just made
top_custs = df.groupby('Customer Name')['Subcat Diff'].mean()
# and sort the values + show the top 10 -- Mitch is a high roller.....
top_custs.sort_values(ascending=False).head(10)

Customer Name
Mitch Willingham        2404.056459
Christopher Martinez    1862.436871
Jim Radford             1628.279903
Andy Reiter             1373.983641
Adrian Barton           1337.826612
Sanjit Chand            1236.837470
Alex Avila              1202.507813
Bill Shonely             880.808891
Sanjit Engle             750.951436
Jenna Caffey             728.799638
Name: Subcat Diff, dtype: float64

In [185]:
df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0         1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
27       28  US-2015-150630 2015-09-17 2015-09-21  Standard Class    TB-21520   
38       39  CA-2015-117415 2015-12-27 2015-12-31  Standard Class    SN-20710   
189     190  CA-2015-102281 2015-10-12 2015-10-14     First Class    MP-17470   
192     193  CA-2015-102281 2015-10-12 2015-10-14     First Class    MP-17470   

       Customer Name      Segment        Country           City  ...  \
0        Claire Gute     Consumer  United States      Henderson  ...   
27   Tracy Blumstein     Consumer  United States   Philadelphia  ...   
38      Steve Nguyen  Home Office  United States        Houston  ...   
189      Mark Packer  Home Office  United States  New York City  ...   
192      Mark Packer  Home Office  United States  New York City  ...   

    Sub-Category                                       Product Name  \
0      Bookcases                  Bush Somerset Collection Bookcase   
27     Bookcases  Riverside Palais Royal Lawyers Bookcase, Royal...   
38     Bookcases  Atlantic Metals Mobile 3-Shelf Bookcases, Cust...   
189    Bookcases  Atlantic Metals Mobile 4-Shelf Bookcases, Cust...   
192    Bookcases  Atlantic Metals Mobile 3-Shelf Bookcases, Cust...   

         Sales Quantity Discount     Profit Cat_Average  Cat Difference  \
0     261.9600        2     0.00    41.9136  349.834887      -25.118960   
27   3083.4300        7     0.50 -1665.0522  349.834887      781.395800   
38    532.3992        3     0.32   -46.9764  349.834887       52.185851   
189   899.1360        4     0.20   112.3920  349.834887      157.017248   
192   626.3520        3     0.20    46.9764  349.834887       79.042178   

     Subcat Avg   Subcat Diff  
0    -15.230509   -375.195009  
27   -15.230509  10832.347861  
38   -15.230509    208.436184  
189  -15.230509   -837.939892  
192  -15.230509   -408.436184  

[5 rows x 25 columns]